java/arcs/core/analysis/ExpressionClaimDeducer.kt

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.analysis
+
+import arcs.core.data.Claim
+import arcs.core.data.expression.Expression
+
+/**
+ * A visitor to help create [Claim]s for Paxel [Expression]s.
+ *
+ * For each [Expression], this visitor produces a [Deduction], which can be translated into a set
+ * of [Claim] relationships.
+ *
+ * [Deduction]s collect [Expression.FieldExpression]s as [Deduction.Equal]s and associate
+ * them with other fields. These are recursive structures, and can represent [Claim] relationships
+ * for deeply nested [Expression]s.
+ */
+class ExpressionClaimDeducer : Expression.Visitor<Deduction, Unit> {
+  override fun <E, T> visit(expr: Expression.UnaryExpression<E, T>, ctx: Unit): Deduction {
+    TODO("Not yet implemented")
+  }
+
+  override fun <L, R, T> visit(expr: Expression.BinaryExpression<L, R, T>, ctx: Unit) =
+    expr.left.accept(this, ctx) + expr.right.accept(this, ctx)
+
+  override fun <T> visit(expr: Expression.FieldExpression<T>, ctx: Unit) =
+    when (val lhs = expr.qualifier) {
```

Comment:
Wouldn't it make more sense to first evaluate `expr.qualifier` and then switch on whether the result is Equal, Derive or Scope?

With equal you append, with scope you lookup, with Derive we don't know what to do, you probably fail.

Reply:
We could make a case for having multiple paths with Equal, which means "result is equal to one of those".

Then in the union case:
```
union(foo, bar).x
```

First, `union(foo, bar)` could evaluate to `Equal(listOf('foo'), listOf('bar'))` and then the fieldExpr would simply append .x to these lists, producing: `Equal(listOf('foo', 'x'), listOf('bar', 'x'))`

Reply:
This is the cartesian cross that Gogul mentioned several comments up :) 

I like this idea -- is it alright if I make the change in a future PR (when we have `union` implemented)?